realm/realm-library/src/main/java/io/realm/Realm.java

```diff
@@ -1154,7 +1160,11 @@ private void checkHasPrimaryKey(Class<? extends RealmObject> clazz) {
     protected void lastLocalInstanceClosed() {
         // validatedRealmFiles must not modified while other thread is executing createAndValidate()
         synchronized (BaseRealm.class) {
-            validatedRealmFiles.remove(configuration.getPath());
+            // All Realm instances with this file have been closed. Decrease the counter.
+            releaseRealmFileReference(configuration);
+            if (typedRealmFileReferenceCounter.get(configuration.getPath()) == null) {
```

Comment:
How about changing `releaseRealmFileReference()` to return counter number.

If so, we can write these two lines into `if (releaseRealmFileReference(configuration) == 0) {`
and reduce one hash map search.